### PR TITLE
add watsonx embeddings provider + add granite chat template + add wat…

### PR DESCRIPTION
**Pull request by mq200**
…sonx models

## Description

- add watsonx embeddings provider
- intelliJ @folder: enable selection of nested folders
- add granite chat template
- add all granite code models
- fix missing watsonx icon in intelliJ plugin
- fix @codebase behavior not respecting configured nFinal/nRetrieve (n being used as num files instead of num chunks for recently edited files)
- align watsonx config options to default options and add apikey URL

## Checklist

- [x] The base branch of this PR is `dev`, rather than `main`
- [x] The relevant docs, if any, have been updated or created

### CODE REVIEW

**Review by Thompson1985 — APPROVED** [review]

File: core/index.d.ts [hunk 1]
```diff
@@ -830,6 +832,13 @@ export interface EmbedOptions {
 
   // AWS and GCP Options
   region?: string;
+
+  // Watsonx Options
+  watsonxUrl?: string;
```
Comment (sestinj) [inline core/index.d.ts]: Can these be turned into a single object? (watsonXConfig)

File: core/llm/index.ts [hunk 2]
```diff
@@ -177,7 +175,7 @@ export abstract class BaseLLM implements ILLM {
       autodetectTemplateFunction(
         options.model,
         this.providerName,
-        options.template,
+        templateType || options.template,
```
Comment (sestinj) [inline core/llm/index.ts]: Let's make this a separate PR—I'd like to be able to separately review so that this doesn't hold up the rest of the code (it's a particularly touchy part of the codebase)

File: extensions/intellij/src/main/kotlin/com/github/continuedev/continueintellijextension/continue/IdeProtocolClient.kt [hunk 3]
```diff
@@ -641,9 +641,14 @@ class IdeProtocolClient(
                     "listFolders" -> {
                         val workspacePath = workspacePath ?: return@launch
                         val workspaceDir = File(workspacePath)
-                        val folders =
-                            workspaceDir.listFiles { file -> file.isDirectory }?.map { file -> file.absolutePath }
-                                ?: emptyList()
+                        val folders = mutableListOf<String>()
```
Comment (sestinj) [inline extensions/intellij/src/main/kotlin/com/github/continuedev/continueintellijextension/continue/IdeProtocolClient.kt:575]: Ideally this would be a separate PR as well, but I'm okay with this one being here, as it looks much safer than the template change I commented on above

**Review by sestinj — COMMENTED** [review]
thanks for all of this!

File: core/llm/index.ts (hunk 2 above)
Comment (mq200) [reply core/llm/index.ts]: I removed this from PR for now- watsonx template names have to be specified in model configs

File: core/index.d.ts (hunk 1 above)
Comment (mq200) [reply core/index.d.ts]: Updated the model config to align fields with default options- now the only field specific to watsonx is the deployment ID (as well as projectId for EmbedOptions)
